docs/user/implement_env.md

```diff
@@ -0,0 +1,308 @@
+# Adding a New Environment
+
+Garage uses the [`Environment`](https://github.com/rlworkgroup/garage/blob/master/src/garage/_environment.py)
+API for all RL environments.
+
+The public API methods of :`Environment` are:
+
+* `reset()`: Resets the environment. Return the first observation, and the episode-level information.
+* `step()`: Steps the environment with an action and returns an EnvStep.
+
+    An `EnvStep` is a named tuple of `env_spec`, `observation`, `action`, `reward`,
+    `step_type` representing a single step in the environment.
+* `render()`: Returns value for rendering, could be a terminal text, an image, etc.
+* `visualize()`: Creates a visualization of the environment.
+* `close()`: Closes the environment.
+
+The attributes of `Environment` are:
+
+* `action_space`: The action space specification.
+* `observation_space`: The observation space specification.
+* `spec`: The environment specifications.
+* `render_modes`: The list of supported render modes.
+
+You should expect to use all environments in garage with this API.
+
+## Add Support for an Existing Environment Library
+
+Luckily, Garage already came with a variety of external environment wrappers:
+
+* `garage.envs.GymEnv`, which supports the `gym` API.
+* `garage.envs.dm_control.DmControlEnv`, which supports Deepmind's
+`dm_control` API.
+* `garage.envs.bullet.BulletEnv`, which supports the `pybullet` API.
+
+    Note that `pybullet` environments are `gym`-based environments, and they
+    can be wrapped with `garage.envs.GymEnv` as well. In such case,
+    `garage.envs.GymEnv` will detect the bullet-based environment and returns
+    a `garage.envs.bullet.BulletEnv`.
+
+If you would like to add an existing environment to garage, you will probably
+need an environment wrapper to handle environment-specific logic.
+
+Here is an handy To-do list for adding a new environment wrapper:
+
+* **Inherit** the environment wrapper from `Environment`.
```

Comment:
This is never required in Python because of duck typing. They merely need to implement the interface methods.